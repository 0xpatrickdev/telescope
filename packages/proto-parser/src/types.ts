
export interface ProtoDep {
    filename: string;
    package: string;
    imports: string[]
}
export interface ProtoRef {
    absolute: string;
    filename: string;
    proto: ProtoRoot;
}

export interface ProtoRoot {
    package: string;
    imports: string[];
    root: any;
};
