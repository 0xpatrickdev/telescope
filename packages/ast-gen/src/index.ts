export * from './types';
export * from './amino';
export * from './client';
export * from './messages';
export * from './utils';