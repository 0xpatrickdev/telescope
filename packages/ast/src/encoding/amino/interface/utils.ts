import * as t from '@babel/types';
import { memberExpressionOrIdentifierAminoCasing, shorthandProperty } from '../../../utils';
import { AminoOptions } from '../types';
import { arrayTypeNDim, BILLION } from '../utils';
import { getTypeFromContext, protoFieldsToArray } from '../utils';
import { getTSTypeFromProtoType, ProtoField } from '../../types';
import { RenderAminoField, renderAminoField } from '.';

export const aminoInterface = {
    defaultType(field: ProtoField, options: AminoOptions, ast: any) {
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(ast)
        );
    },
    long(field: ProtoField, options: AminoOptions) {
        // longs become strings...
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(t.tSStringKeyword())
        )
    },
    height(field: ProtoField, options: AminoOptions) {
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(
                t.tsTypeReference(t.identifier('AminoHeight'))
            )
        );
    },
    enum(field: ProtoField, options: AminoOptions) {
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(t.tSNumberKeyword())
        );
    },
    enumArray(field: ProtoField, options: AminoOptions) {
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(arrayTypeNDim(t.tSNumberKeyword(), 1))
        );
    },
    type({ context, field, options }: RenderAminoField) {
        const Type = getTypeFromContext(field, context);
        const properties = protoFieldsToArray(Type).map(field => {
            return renderAminoField({
                context,
                field,
                options
            })
        });

        // 
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(
                t.tsTypeLiteral(
                    properties
                )
            )
        );
    },
    typeArray({ context, field, options }: RenderAminoField) {
        const Type = getTypeFromContext(field, context);
        const properties = protoFieldsToArray(Type).map(field => {
            return renderAminoField({
                context,
                field,
                options
            })
        });

        // 
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(
                arrayTypeNDim(t.tsTypeLiteral(
                    properties
                ), 1)
            )
        );
    },

    array(field: ProtoField, options: AminoOptions) {
        // TODO write test case 

        // return t.tsPropertySignature(
        //     t.identifier(options.aminoCasingFn(field.name)),
        //     t.tsTypeAnnotation(
        //         arrayTypeNDim(t.tsTypeLiteral(
        //             properties
        //         ), 1)
        //     )
        // );
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(
                arrayTypeNDim(
                    getTSTypeFromProtoType(field.type),
                    1
                )
            )
        );
    }
}