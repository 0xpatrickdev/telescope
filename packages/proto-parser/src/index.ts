export * from './store';
export * from './resolver';
export * from './utils';
export * from './types';
export * from './traverse';
export * from './lookup';