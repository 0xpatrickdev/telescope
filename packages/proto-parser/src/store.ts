import { sync as glob } from 'glob';
import { parse } from 'protobufjs';
import { readFileSync } from 'fs';
import { join, resolve as pathResolve } from 'path';
import { ProtoDep, ProtoRef } from './types';
import { getNestedProto } from './utils';

export class ProtoStore {
    files: string[];
    protoDir: string;
    deps: ProtoDep[];
    protos: any[];

    constructor(protoDir) {
        this.protoDir = pathResolve(protoDir);
    }

    findProto(filename): ProtoRef {
        return this.getProtos().find(proto => {
            return proto.filename === filename
        });
    }

    findProtoObject(filename, name): any {
        const proto = this.findProto(filename);
        return getNestedProto(proto.proto)[name];
    }

    getProtos(): ProtoRef[] {
        if (this.protos) return this.protos;
        const protoSplat = join(this.protoDir, '/**/*.proto');
        const protoFiles = glob(protoSplat);
        const contents = protoFiles.map(filename => ({
            absolute: filename,
            filename: filename.split(this.protoDir)[1].replace(/^\//, ''),
            content: readFileSync(filename, 'utf-8')
        }))
        const protos = contents.map(({ absolute, filename, content }) => {
            return {
                proto: parse(content, {
                    keepCase: false,
                    alternateCommentMode: true,
                    preferTrailingComment: false
                }), absolute, filename
            };
        });
        this.protos = protos;
        return protos;
    }

    getDeps(): ProtoDep[] {
        if (this.deps) return this.deps;
        const deps = this.getProtos().map(el => {
            const {
                filename,
                proto: {
                    package: pkg,
                    imports
                }
            } = el;
            return {
                filename,
                pkg,
                imports
            }
        });
        this.deps = deps;
        return deps;
    }

}
