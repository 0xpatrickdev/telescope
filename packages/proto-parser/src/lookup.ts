import { ProtoStore } from './store';
import { ProtoRef, ProtoRoot } from './types';
import { getNestedProto } from './utils';

export const recursiveLookup = (proto: any, name: string, scope: string[] = [], allowNested = true) => {
    if (!proto) return;

    if (proto.hasOwnProperty(name)) return {
        name,
        scope,
        ...proto[name]
    }

    if (allowNested && proto) {
        const keys = Object.keys(proto);
        for (let k = 0; k < keys.length; k++) {
            const found = recursiveLookup(proto[keys[k]].nested, name, [...scope, keys[k]], allowNested);
            if (found) {
                return {
                    name,
                    ...found
                };
            };
        }
    }
};

/*

  "imports": [
    "google/api/expr/v1alpha1/syntax.proto",
    "google/protobuf/empty.proto",
    "google/protobuf/struct.proto"
  ],

  ...

  "type": "Empty"

  finds Empty inside of the import

*/
export const importLookup = (store: ProtoStore, root: ProtoRoot, name: string) => {
    const objectsFromImports = root.imports.map(imp => {
        const ref = store.findProto(imp);
        return {
            name,
            import: imp,
            importedName: name,
            obj: lookup(store, ref.proto, name, false),
        };
    }).filter(a => !!a.obj);
    if (objectsFromImports.length) return objectsFromImports[0];
};


/*

  "imports": [
    "google/api/expr/v1alpha1/syntax.proto",
    "google/protobuf/empty.proto",
    "google/protobuf/struct.proto"
  ],

  ...

  "type": "google.protobuf.NullValue"

  finds NullValue directly from package name

*/

export const protoImportLookup = (store: ProtoStore, root: ProtoRoot, name: string) => {
    if (name.startsWith('.')) name = name.replace(/^\./, '');
    const nameAsArray = name.split('.');
    const objectName = nameAsArray.pop();
    const packageName = nameAsArray.join('.')
    const objs = store.getProtos()
        // 1 ask all package files of package, e.g "google.protobuf"
        .filter(proto => proto.proto.package === packageName)
        // 2 cross-checking w the imports by filter()
        .filter(proto => root.imports.includes(proto.filename))
        .map((ref: ProtoRef) => {
            return {
                import: ref.filename,
                obj: lookup(store, ref.proto, objectName, false)
            }
        })
        .filter(a => !!a.obj);
    if (objs.length) {
        return {
            import: objs[0].import,
            obj: objs[0].obj,
            importedName: name,
            name: objectName,
            package: packageName
        }
    }
};

export const lookup = (store: ProtoStore, root: ProtoRoot, name: string, allowNested = true) => {
    const nested = getNestedProto(root);
    return recursiveLookup(nested, name, [root.package], allowNested);
};
