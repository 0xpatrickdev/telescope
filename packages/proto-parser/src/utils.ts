import dotty from 'dotty';
import { Service, Type, Enum, Root, Namespace } from 'protobufjs';
import { ProtoStore } from './store';
import { ProtoRef, ProtoRoot } from './types';

export const getNestedProto = (root: ProtoRoot) => {
    const nestedPath = 'root.nested.' + root.package.split('.').join('.nested.') + '.nested';
    return dotty.get(root, nestedPath);
};

export const getServices = (root: ProtoRoot) => {
    const nested = getNestedProto(root);
    return Object.keys(nested).map(key => {
        return nested[key];
    }).filter(obj => {
        return obj instanceof Service;
    }).map(el => {
        return {
            name: el.name,
            ...el.toJSON({ keepComments: true })
        };
    });
};

export const getTypes = (root: ProtoRoot) => {
    const nested = getNestedProto(root);
    return Object.keys(nested).map(key => {
        return nested[key];
    }).filter(obj => {
        return obj instanceof Type;
    }).map(el => {
        return {
            name: el.name,
            ...el.toJSON({ keepComments: true })
        };
    });
};

export const getEnums = (root: ProtoRoot) => {
    const nested = getNestedProto(root);
    return Object.keys(nested).map(key => {
        return nested[key];
    }).filter(obj => {
        return obj instanceof Enum;
    }).map(el => {
        return {
            name: el.name,
            ...el.toJSON({ keepComments: true })
        };
    });
};

export const recursiveLookup = (proto: any, name: string, scope: string[] = [], allowNested = true) => {
    if (!proto) return;

    if (proto.hasOwnProperty(name)) return {
        name,
        scope,
        ...proto[name]
    }

    if (allowNested && proto) {
        const keys = Object.keys(proto);
        for (let k = 0; k < keys.length; k++) {
            const found = recursiveLookup(proto[keys[k]].nested, name, [...scope, keys[k]], allowNested);
            if (found) {
                return {
                    name,
                    ...found
                };
            };
        }
    }
};

/*

  "imports": [
    "google/api/expr/v1alpha1/syntax.proto",
    "google/protobuf/empty.proto",
    "google/protobuf/struct.proto"
  ],

  ...

  "type": "Empty"

  finds Empty inside of the import

*/
export const importLookup = (store: ProtoStore, root: ProtoRoot, name: string) => {
    const objectsFromImports = root.imports.map(imp => {
        const ref = store.findProto(imp);
        return {
            name,
            import: imp,
            importedName: name,
            obj: lookup(store, ref.proto, name, false),
        };
    }).filter(a => !!a.obj);
    if (objectsFromImports.length) return objectsFromImports[0];
};


/*

  "imports": [
    "google/api/expr/v1alpha1/syntax.proto",
    "google/protobuf/empty.proto",
    "google/protobuf/struct.proto"
  ],

  ...

  "type": "google.protobuf.NullValue"

  finds NullValue directly from package name

*/

export const protoImportLookup = (store: ProtoStore, root: ProtoRoot, name: string) => {
    if (name.startsWith('.')) name = name.replace(/^\./, '');
    const nameAsArray = name.split('.');
    const objectName = nameAsArray.pop();
    const packageName = nameAsArray.join('.')
    const objs = store.getProtos()
        // 1 ask all package files of package, e.g "google.protobuf"
        .filter(proto => proto.proto.package === packageName)
        // 2 cross-checking w the imports by filter()
        .filter(proto => root.imports.includes(proto.filename))
        .map((ref: ProtoRef) => {
            return {
                import: ref.filename,
                obj: lookup(store, ref.proto, objectName, false)
            }
        })
        .filter(a => !!a.obj);
    if (objs.length) {
        return {
            import: objs[0].import,
            obj: objs[0].obj,
            importedName: name,
            name: objectName,
            package: packageName
        }
    }
};

export const lookup = (store: ProtoStore, root: ProtoRoot, name: string, allowNested = true) => {
    const nested = getNestedProto(root);
    return recursiveLookup(nested, name, [root.package], allowNested);
};

/*
    nested objects get a slightly different naming convention
    e.g. SignatureDescriptor_Data or SignatureDescriptor_Data_Multi
*/

export const getObjectName = (name: string, scope: string[] = []) => {
    if (!scope.length || scope.length === 1) return name;
    const [_pkg, ...scopes] = scope;
    return [...scopes, name].join('_')
};


export const traverse = (store: ProtoStore, root: ProtoRoot) => {
    const imports = {};

    const obj: ProtoRoot & { parsedImports: Record<string, string[]> } = {
        imports: root.imports,
        package: root.package,
        root: recursiveTraversal(store, root, root.root, imports),
        parsedImports: null
    };
    obj.parsedImports = imports;
    return obj;
};


const NATIVE_TYPES = [
    "double",   // 0
    "float",    // 1
    "int32",    // 2
    "uint32",   // 3
    "sint32",   // 4
    "fixed32",  // 5
    "sfixed32", // 6
    "int64",    // 7
    "uint64",   // 8
    "sint64",   // 9
    "fixed64",  // 10
    "sfixed64", // 11
    "bool",     // 12
    "string",   // 13
    "bytes"     // 14
];


const parseFields = (store: ProtoStore, root: ProtoRoot, obj: any, imports: object) => {
    return Object.keys(obj.fields).reduce((m, key) => {

        const field = obj.fields[key];
        let found: any = null;

        if (NATIVE_TYPES.includes(field.type)) {
            m[key] = {
                parsedType: { name: field.type, type: 'native' },
                isScalar: true,
                typeNum: NATIVE_TYPES.indexOf(field.type),
                ...field.toJSON({ keepComments: true })
            }
            return m;
        }

        found = importLookup(store, root, field.type);
        if (found) {
            imports[found.import] = imports[found.import] || [];
            imports[found.import] = [...new Set([...imports[found.import], found.name])];
            m[key] = {
                parsedType: instanceType(found.obj),
                scopeType: 'import',
                scope: [found.obj.scope],
                ...field.toJSON({ keepComments: true }),
                importedName: found.importedName,
                import: found.import,
            };
            return m;
        }

        found = protoImportLookup(store, root, field.type);
        if (found) {
            imports[found.import] = imports[found.import] || [];
            imports[found.import] = [...new Set([...imports[found.import], found.name])];
            m[key] = {

                parsedType: instanceType(found.obj),
                scopeType: 'protoImport',
                scope: [found.package],
                ...field.toJSON({ keepComments: true }),
                importedName: found.importedName,
                import: found.import,
            };
            return m;
        }

        found = lookup(store, root, field.type);
        if (found) {
            m[key] = {
                scope: found.scope,
                parsedType: instanceType(found),
                ...field.toJSON({ keepComments: true }),
            };
            return m;
        }

        return m;
    }, {});
};

const parseType = (store: ProtoStore, root: ProtoRoot, obj: any, imports: object) => {
    let nested = null;
    if (obj.nested) {
        nested = Object.keys(obj.nested).reduce((m, key) => {
            m[key] = recursiveTraversal(store, root, obj.nested[key], imports);
            return m;
        }, {});
    }
    return {
        type: 'Type',
        name: obj.name,
        options: obj.options,
        fields: parseFields(store, root, obj, imports),
        nested
    }
};

const parseEnum = (store: ProtoStore, root: ProtoRoot, obj: any, imports: object) => {
    return {
        type: 'Enum',
        name: obj.name,
        ...obj.toJSON({ keepComments: true })
    }
};

const parseService = (store: ProtoStore, root: ProtoRoot, obj: any, imports: object) => {
    let nested = null;
    if (obj.nested) {
        nested = Object.keys(obj.nested).reduce((m, key) => {
            m[key] = recursiveTraversal(store, root, obj.nested[key], imports);
            return m;
        }, {});
    }

    return {
        type: 'Service',
        ...obj.toJSON({ keepComments: true }),
        nested
    }
};

export const recursiveTraversal = (store: ProtoStore, root: ProtoRoot, obj: any, imports: object) => {
    if (obj instanceof Type) {
        return parseType(store, root, obj, imports);
    }
    if (obj instanceof Enum) {
        return parseEnum(store, root, obj, imports);
    }
    if (obj instanceof Service) {
        return parseService(store, root, obj, imports);
    }
    if (obj instanceof Root) {
        if (obj.nested) {
            return Object.keys(obj.nested).reduce((m, key) => {
                m.nested[key] = recursiveTraversal(store, root, obj.nested[key], imports);
                return m;
            }, {
                type: 'Root',
                nested: {}
            });
        } else {
            throw new Error('recursiveTraversal() cannot find protobufjs Type')
        }
    }
    if (obj instanceof Namespace) {
        if (obj.nested) {
            return Object.keys(obj.nested).reduce((m, key) => {
                m.nested[key] = recursiveTraversal(store, root, obj.nested[key], imports);
                return m;
            }, {
                type: 'Namespace',
                nested: {}
            });
        } else {
            throw new Error('recursiveTraversal() cannot find protobufjs Type')
        }
    }
    throw new Error('recursiveTraversal() cannot find protobufjs Type')
};

export const instanceType = (obj: any) => {
    if (obj instanceof Type) {
        return {
            name: obj.name,
            type: 'Type'
        };
    }
    if (obj instanceof Enum) {
        return {
            name: obj.name,
            type: 'Enum'
        };
    }
    if (obj instanceof Service) {
        return {
            name: obj.name,
            type: 'Service'
        };
    }
    if (obj instanceof Root) {
        return {
            type: 'Root'
        };
    }
    if (obj instanceof Namespace) {
        return {
            type: 'Namespace'
        };
    }
    // if (obj.name === 'Timestamp') {
    //     return {
    //         name: obj.name,
    //         type: 'google'
    //     };
    // }
    // if (obj.name === 'Duration') {
    //     return {
    //         name: obj.name,
    //         type: 'google'
    //     };
    // }
    if (obj.name.match(/^[a-z]/)) {
        throw new Error('instanceType() cannot find protobufjs Type')
    }
    // duck typing... 
    // TODO why did we loose instance tyeps/names?
    if (obj.fields) {
        return {
            name: obj.name,
            type: 'Type'
        };
    }
    if (obj.values) {
        return {
            name: obj.name,
            type: 'Enum'
        };
    }
    throw new Error('instanceType() cannot find protobufjs Type')
};
