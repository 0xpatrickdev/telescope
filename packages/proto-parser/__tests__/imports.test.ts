import { ProtoResolver, ProtoStore } from '../src/index'
import { importLookup, lookup, protoImportLookup } from '../src/utils';
const store = new ProtoStore(__dirname + '/../__fixtures__/chain1');

describe('imports lookup', () => {
    it('Any', () => {
        // - [x] google.protobuf.Any -> Any (imported)
        const ref = store.findProto('cosmos/app/v1alpha1/config.proto');
        // https://github.com/pyramation/protobuf-example-reference/blob/master/json/cosmos/app/v1alpha1/config.json#L31
        // https://github.com/pyramation/protobuf-example-reference/blob/master/ts/cosmos/app/v1alpha1/config.ts#L4
        const { obj: _obj, ...Any } = importLookup(store, ref.proto, 'Any');
        expect(Any).toMatchSnapshot();
    });
    describe('google/api/expr/v1alpha1/checked', () => {
        const ref = store.findProto('google/api/expr/v1alpha1/checked.proto');
        // - [x] Empty -> Empty (imported)
        // - [x] NullValue -> NullValue (imported enum)
        // - [x] google.protobuf.Empty -> Empty (imported)
        // - [x] google.protobuf.NullValue -> NullValue (imported enum)
        // https://github.com/pyramation/protobuf-example-reference/blob/master/json/google/api/expr/v1alpha1/checked.json#L74-L78
        // https://github.com/pyramation/protobuf-example-reference/blob/master/ts/google/api/expr/v1alpha1/checked.ts#L5-L6        
        it('Empty', () => {
            const { obj, ...Empty } = importLookup(store, ref.proto, 'Empty');
            expect(obj.name).toEqual('Empty')
            expect(obj.scope).toEqual(['google.protobuf'])
            expect(Empty).toMatchSnapshot();
        });
        it('NullValue', () => {
            const { obj, ...NullValue } = importLookup(store, ref.proto, 'NullValue');
            expect(obj.name).toEqual('NullValue')
            expect(obj.scope).toEqual(['google.protobuf'])
            expect(NullValue).toMatchSnapshot();
        });
        it('google.protobuf.NullValue', () => {
            const { obj, importedName, ...NullValue } = protoImportLookup(store, ref.proto, 'google.protobuf.NullValue');
            expect(obj.name).toEqual('NullValue')
            expect(obj.scope).toEqual(['google.protobuf'])
            expect(importedName).toEqual('google.protobuf.NullValue')
            expect(NullValue).toMatchSnapshot();
        });
        it('google.protobuf.Empty', () => {
            const { obj, importedName, ...Empty } = protoImportLookup(store, ref.proto, 'google.protobuf.Empty');
            expect(obj.name).toEqual('Empty')
            expect(importedName).toEqual('google.protobuf.Empty')
            expect(obj.scope).toEqual(['google.protobuf'])
            expect(Empty).toMatchSnapshot();
        });
    });
    describe('osmosis/claim/v1beta1/params.proto', () => {
        const ref = store.findProto('osmosis/claim/v1beta1/params.proto');
        // - [x] google.protobuf.Duration -> string | Duration (imported from google)
        // https://github.com/pyramation/protobuf-example-reference/blob/master/json/osmosis/claim/v1beta1/params.json#L32-L38
        // https://github.com/pyramation/protobuf-example-reference/blob/master/ts/osmosis/claim/v1beta1/params.ts#L5    });
        it('google.protobuf.Duration', () => {
            const { obj, importedName, ...Duration } = protoImportLookup(store, ref.proto, 'google.protobuf.Duration');
            expect(obj.name).toEqual('Duration')
            expect(importedName).toEqual('google.protobuf.Duration')
            expect(obj.scope).toEqual(['google.protobuf'])
            expect(Duration).toMatchSnapshot();
        });
        it('Duration', () => {
            const { obj, ...Duration } = importLookup(store, ref.proto, 'Duration');
            expect(obj.name).toEqual('Duration')
            expect(obj.scope).toEqual(['google.protobuf'])
            expect(Duration).toMatchSnapshot();
        });
    });
    it('cosmos.base.query.v1beta1.PageResponse', () => {
        const ref = store.findProto('cosmos/authz/v1beta1/query.proto');
        // - [x] cosmos.base.query.v1beta1.PageResponse -> PageResponse (imported)
        // https://github.com/pyramation/protobuf-example-reference/blob/master/json/cosmos/authz/v1beta1/query.json#L101
        // https://github.com/pyramation/protobuf-example-reference/blob/master/ts/cosmos/authz/v1beta1/query.ts#L27
        const { obj, importedName, ...PageResponse } = protoImportLookup(store, ref.proto, 'cosmos.base.query.v1beta1.PageResponse');
        expect(obj.name).toEqual('PageResponse')
        expect(importedName).toEqual('cosmos.base.query.v1beta1.PageResponse')
        expect(obj.scope).toEqual(['cosmos.base.query.v1beta1'])
        expect(PageResponse).toMatchSnapshot();

    });
    describe('cosmos/tx/v1beta1/service.proto', () => {
        const ref = store.findProto('cosmos/tx/v1beta1/service.proto');
        it('.tendermint.types.BlockID', () => {
            //  - [x] .tendermint.types.BlockID -> BlockID (imported from tendermint/types/types)
            //  - [x] .tendermint.types.Block ->Block (imported from tendermint/types/block)
            // https://github.com/pyramation/protobuf-example-reference/blob/master/ts/cosmos/tx/v1beta1/service.ts#L7-L8
            // https://github.com/pyramation/protobuf-example-reference/blob/master/json/cosmos/tx/v1beta1/service.json#L238-L242
            const { obj, importedName, ...BlockID } = protoImportLookup(store, ref.proto, '.tendermint.types.BlockID');
            expect(obj.name).toEqual('BlockID')
            expect(importedName).toEqual('tendermint.types.BlockID')
            expect(obj.scope).toEqual(['tendermint.types'])
            expect(BlockID).toMatchSnapshot();
        });
        it('.tendermint.types.Block', () => {
            //  - [x] .tendermint.types.BlockID -> BlockID (imported from tendermint/types/types)
            //  - [x] .tendermint.types.Block ->Block (imported from tendermint/types/block)
            // https://github.com/pyramation/protobuf-example-reference/blob/master/ts/cosmos/tx/v1beta1/service.ts#L7-L8
            // https://github.com/pyramation/protobuf-example-reference/blob/master/json/cosmos/tx/v1beta1/service.json#L238-L242
            const { obj, importedName, ...Block } = protoImportLookup(store, ref.proto, '.tendermint.types.Block');
            expect(obj.name).toEqual('Block')
            expect(importedName).toEqual('tendermint.types.Block')
            expect(obj.scope).toEqual(['tendermint.types'])
            expect(Block).toMatchSnapshot();
        });
    })
    it('google.rpc.Status', () => {
        // - [x] google.rpc.Status -> Status (imported from status rpc nested Status obj)
        // https://github.com/pyramation/protobuf-example-reference/blob/master/json/google/api/expr/v1alpha1/eval.json#L82
        // https://github.com/pyramation/protobuf-example-reference/blob/master/json/google/rpc/status.json#L21
        const ref = store.findProto('google/api/expr/v1alpha1/eval.proto');
        const { obj, importedName, ...Status } = protoImportLookup(store, ref.proto, 'google.rpc.Status');
        expect(obj.name).toEqual('Status')
        expect(importedName).toEqual('google.rpc.Status')
        expect(obj.scope).toEqual(['google.rpc'])
        expect(Status).toMatchSnapshot();
    });
    it('ibc.core.client.v1.Height', () => {
        // - [x] ibc.core.client.v1.Height -> Height (imported from ibc/core/client/v1/client)
        // https://github.com/pyramation/protobuf-example-reference/blob/master/json/ibc/core/channel/v1/query.json#L234
        // https://github.com/pyramation/protobuf-example-reference/blob/master/ts/ibc/core/channel/v1/query.ts#L5
        const ref = store.findProto('ibc/core/channel/v1/query.proto');
        const { obj, importedName, ...Height } = protoImportLookup(store, ref.proto, 'ibc.core.client.v1.Height');
        expect(obj.name).toEqual('Height');
        expect(importedName).toEqual('ibc.core.client.v1.Height')
        expect(obj.scope).toEqual(['ibc.core.client.v1'])
        expect(Height).toMatchSnapshot();
    });
});