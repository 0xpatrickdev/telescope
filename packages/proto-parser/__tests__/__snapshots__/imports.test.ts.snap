// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`imports lookup Any 1`] = `
Object {
  "import": "google/protobuf/any.proto",
  "name": "Any",
}
`;

exports[`imports lookup cosmos.base.query.v1beta1.PageResponse 1`] = `
Object {
  "import": "cosmos/base/query/v1beta1/pagination.proto",
  "name": "PageResponse",
  "package": "cosmos.base.query.v1beta1",
}
`;

exports[`imports lookup cosmos/tx/v1beta1/service.proto .tendermint.types.Block 1`] = `
Object {
  "import": "tendermint/types/block.proto",
  "name": "Block",
  "package": "tendermint.types",
}
`;

exports[`imports lookup cosmos/tx/v1beta1/service.proto .tendermint.types.BlockID 1`] = `
Object {
  "import": "tendermint/types/types.proto",
  "name": "BlockID",
  "package": "tendermint.types",
}
`;

exports[`imports lookup google.rpc.Status 1`] = `
Object {
  "import": "google/rpc/status.proto",
  "name": "Status",
  "package": "google.rpc",
}
`;

exports[`imports lookup google/api/expr/v1alpha1/checked Empty 1`] = `
Object {
  "import": "google/protobuf/empty.proto",
  "name": "Empty",
}
`;

exports[`imports lookup google/api/expr/v1alpha1/checked NullValue 1`] = `
Object {
  "import": "google/protobuf/struct.proto",
  "name": "NullValue",
}
`;

exports[`imports lookup google/api/expr/v1alpha1/checked google.protobuf.Empty 1`] = `
Object {
  "import": "google/protobuf/empty.proto",
  "name": "Empty",
  "package": "google.protobuf",
}
`;

exports[`imports lookup google/api/expr/v1alpha1/checked google.protobuf.NullValue 1`] = `
Object {
  "import": "google/protobuf/struct.proto",
  "name": "NullValue",
  "package": "google.protobuf",
}
`;

exports[`imports lookup ibc.core.client.v1.Height 1`] = `
Object {
  "import": "ibc/core/client/v1/client.proto",
  "name": "Height",
  "package": "ibc.core.client.v1",
}
`;

exports[`imports lookup osmosis/claim/v1beta1/params.proto Duration 1`] = `
Object {
  "import": "google/protobuf/duration.proto",
  "name": "Duration",
}
`;

exports[`imports lookup osmosis/claim/v1beta1/params.proto google.protobuf.Duration 1`] = `
Object {
  "import": "google/protobuf/duration.proto",
  "name": "Duration",
  "package": "google.protobuf",
}
`;
